Changing the logic to identify dotnet isolated apps

soninaren opened the pull request:
<!-- Please provide all the information below.  -->

The original logic for identifying the dotnet isolated apps was using worker config. However, we were getting a false positive on the check if the customer did not have `FUNCTIONS_WORKER_RUNTIME` set. The new check would take into account function metadata and worker runtime instead of worker config. 

### Pull request checklist

* [x] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **should not** be added to the release notes for the next release
    * [ ] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)

<!-- Optional: delete if not applicable  -->
### Additional information

Additional PR information

soninaren committed:
Changing the logic to identify dotnet isolated apps

jviau reviewed (commented):
> [x] I have added all required tests (Unit tests, E2E tests)

Is existing test coverage sufficient? I would feel a bit better if we add a unit test which fails before this change and succeeds afterwards.

soninaren commented:
> > [x] I have added all required tests (Unit tests, E2E tests)
> 
> Is existing test coverage sufficient? I would feel a bit better if we add a unit test which fails before this change and succeeds afterwards.

To be honest the existing unit tests does not cover the missing worker runtime scenario. I am working towards adding the unit / e2e test. The test on this one is going to be complicated. I needed eyes on the change before it makes it to hotfix.

soninaren committed:
Fixing null reference

jviau commented:
> > > [x] I have added all required tests (Unit tests, E2E tests)
> > 
> > 
> > Is existing test coverage sufficient? I would feel a bit better if we add a unit test which fails before this change and succeeds afterwards.
> 
> To be honest the existing unit tests does not cover the missing worker runtime scenario. I am working towards adding the unit / e2e test. The test on this one is going to be complicated. I needed eyes on the change before it makes it to hotfix.

Is there a test plan for this then before we include it in the hotfix?

liliankasem commented:
> > > > [x] I have added all required tests (Unit tests, E2E tests)
> > > 
> > > 
> > > Is existing test coverage sufficient? I would feel a bit better if we add a unit test which fails before this change and succeeds afterwards.
> > 
> > 
> > To be honest the existing unit tests does not cover the missing worker runtime scenario. I am working towards adding the unit / e2e test. The test on this one is going to be complicated. I needed eyes on the change before it makes it to hotfix.
> 
> Is there a test plan for this then before we include it in the hotfix?

Yeah I'm concerned that the hotfix branch was just merged without reviews from our team

soninaren committed:
adding unit tests

soninaren commented:
/azp run

soninaren commented:
> > > > > [x] I have added all required tests (Unit tests, E2E tests)
> > > > 
> > > > 
> > > > Is existing test coverage sufficient? I would feel a bit better if we add a unit test which fails before this change and succeeds afterwards.
> > > 
> > > 
> > > To be honest the existing unit tests does not cover the missing worker runtime scenario. I am working towards adding the unit / e2e test. The test on this one is going to be complicated. I needed eyes on the change before it makes it to hotfix.
> > 
> > 
> > Is there a test plan for this then before we include it in the hotfix?
> 
> Yeah I'm concerned that the hotfix branch was just merged without reviews from our team

Ok, I have added unit tests. We have sufficient coverage here now. 
@liliankasem , @jviau Can you take a look again?

brettsam commented on src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs:
```diff
@@ -101,9 +102,10 @@ public async Task<IEnumerable<Type>> GetExtensionsStartupTypesAsync()
                     isPrecompiledFunctionApp = isPrecompiledFunctionApp || functionMetadata.Language == DotNetScriptTypes.DotNetAssembly;
                 }
+
+                isDotnetIsolatedApp = IsDotnetIsolatedApp(functionMetadataCollection, SystemEnvironment.Instance);
```
Can we use the `IEnvironment` abstraction here? Rather than hardcoding `SystemEnvironment`?

brettsam replied:
Also, why did this check move inside the `bundleConfigured` check?

soninaren replied:
I needed the `functionMetadataCollection` which is only available inside the if block . `isDotnetIsolatedApp` is only used if bundle is configured same as `isPrecompiledFunctionApp` so moving it inside would not break the logic.

soninaren replied:
I don't think IEnvironment abstraction is available to `ScriptStartupTypeLocator`

brettsam replied:
This reads a little confusingly b/c if `!bundlesConfigured`... then `isDotnetIsolatedApp` will always be `false` since we never evaluate it, which isn't really the case.

I think it's the variable naming that's throwing me off. 

Can you change that name to something like `workerSupportsBundles`? That may make things read better... and it'd basically evaluate with this...

` workerSupportsBundles = !IsDotnetIsolatedApp(functionMetadataCollection, SystemEnvironment.Instance);`

soninaren replied:
sure

soninaren replied:
Discussed this with @brettsam offline. Going to merge this PR and follow with the above suggested changes in another PR.

brettsam commented on src/WebJobs.Script/DependencyInjection/ScriptStartupTypeLocator.cs:
```diff
@@ -335,14 +337,10 @@ void CollectError(Type extensionType, Version minimumVersion, ExtensionStartupTy
         {
-            if (workerConfigs != null)
-            {
-                return workerConfigs.Where(config => config.Description is not null && config.Description.Language == RpcWorkerConstants.DotNetIsolatedLanguageWorkerName).Any();
```
So were we previously checking whether there was any dotnet-isolated worker config? and if so, assuming it was dotnet-isolated? 

It seems like this would have broken once we introduced the placeholder worker, which meant this config was now always present. 

Any idea why this started showing up as an issue now?

soninaren replied:
This would have been an issue since it was introduced in 4.26.0. I think we just detected this late.
